File: 0367-valid-perfect-square/0367-valid-perfect-square.ts
function isPerfectSquare(num: number): boolean {
    if(num == 1) return true;
    
    let rightPtr: number = Math.floor(num / 2);
    let leftPtr: number = 1;
    
    while(leftPtr <= rightPtr) {
        let mid: number = Math.floor((rightPtr + leftPtr) / 2);
        
        if(mid * mid == num) {
            return true;
        } else if(mid * mid > num) {
            rightPtr = mid - 1;
        } else {
            leftPtr = mid + 1;
        }
    }
    
    return false;
};

File: 0367-valid-perfect-square/0367-valid-perfect-square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        if num == 1:
            return True
        
        high = num // 2
        low = 1
        
        while low <= high:
            mid = low + (high - low) // 2
            
            if mid * mid == num:
                return True
            elif mid * mid < num:
                low = mid + 1
            else:
                high = mid - 1
                
        return False